paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java

```diff
@@ -290,19 +297,151 @@ public void createTag(String tagName, long fromSnapshotId) {
                 fromSnapshotId);
 
         Snapshot snapshot = snapshotManager.snapshot(fromSnapshotId);
-        TagManager tagManager = new TagManager(fileIO, path);
-        tagManager.createTag(snapshot, tagName);
+        tagManager().createTag(snapshot, tagName);
     }
 
     @Override
     public void deleteTag(String tagName) {
-        checkArgument(!StringUtils.isBlank(tagName), "Tag name '%s' is blank.", tagName);
+        tagManager().deleteTag(tagName, store().newTagDeletion(), snapshotManager());
+    }
+
+    @Override
+    public void rollbackTo(String tagName) {
+        TagManager tagManager = tagManager();
+        checkArgument(tagManager.tagExists(tagName), "Rollback tag '%s' doesn't exist.", tagName);
 
-        TagManager tagManager = new TagManager(fileIO, path);
         Snapshot taggedSnapshot = tagManager.taggedSnapshot(tagName);
+        cleanLargerThan(taggedSnapshot);
+
+        try {
+            // it is possible that the earliest snapshot is later than the rollback tag because of
+            // snapshot expiration, in this case the `cleanLargerThan` method will delete all
+            // snapshots, so we should write the tag file to snapshot directory and modify the
+            // earliest hint
+            SnapshotManager snapshotManager = snapshotManager();
+            if (!snapshotManager.snapshotExists(taggedSnapshot.id())) {
+                fileIO.writeFileUtf8(
+                        snapshotManager().snapshotPath(taggedSnapshot.id()),
+                        fileIO.readFileUtf8(tagManager.tagPath(tagName)));
+                snapshotManager.commitEarliestHint(taggedSnapshot.id());
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Clean snapshots and tags whose id is larger than given snapshot's in the reverse direction.
+     *
+     * <p>Snapshots are cleaned first because it is easier to collect their deletion skipping set.
+     */
+    private void cleanLargerThan(Snapshot snapshot) {
```

Comment:
Create a separate class to contain `cleanLargerThan` method. For example, a `RollbackHelper` class.